src/createSlice.ts

```diff
@@ -35,14 +35,19 @@ export interface Slice<
    * reducer.
    */
   actions: ActionCreators
+
+  caseReducers: SliceCaseReducers<State, any>
```

Comment:
This is too generic, it needs some relation to the original reducers or it will broaden to accept any type of reducer with just any key. I'll add two example tests below.

Reply:
Trying to poke at this now.  Got any suggestions for approaches?

Reply:
A possible solution (that is potentially breaking) is to change the type of the second generic argument on most things in this file; a lot of things here are overly broad, not just the changes done in this PR.

Make them all use the `PA extends PayloadActions` instead of types that are derived from it; put the derivation in the parameter list / helper types / interface declarations themselves. The fundamental units here are the `State` object and the `PayloadActions` that are handled.

Speaking of which, `PayloadActions` itself is erasing types; it does not preserve the "individuality" of each `PayloadAction` it contains; fixing that is even more involved than this,  though, and I think the current problems we're pointing out can be fixed before addressing that (higher kinded types yay 😓).

For example, this `Slice` interface should receive `State` and `PA extends PayloadActions`, the type of `actions` should be (what currently is) `CaseReducerActions<SliceCaseReducers<State, PA>>` (they're all already `createAction`ed, you can no longer have `CaseReducerWithPrepare` objects in the result), and type of of `caseReducers` should be `SliceCaseReducers<State, PA>`.

Reply:
Thanks for the suggestions.

I will freely admit that part of me wants to throw my hands up and say "someone else fix this!", and it may come to that.  But, I'd also like to start becoming at least not completely incompetent in this area.

For the moment, is there a simpler approach that would help improve the current PR without trying to redo the rest of the types right now?  Should we just get this PR in and tackle everything else in a follow-up?

Reply:
For a simple approach:

* the createSlice method signature passes `CaseReducers` on to Slice instead of currently `CaseReducerActions<CaseReducers>`
* Slice changes it's second generic argument from `ActionCreators extends { [key: string]: any } = { [key: string]: any }`to `CaseReducers extends SliceCaseReducerDefinitions<State, unknown>` (I believe the default value would even satisfy that, as broad as it is)
* `actions: ActionCreators` becomes type `CaseReducerActions<CaseReducers>`

that way, you preserve given behaviour, but within Slice, you now have access to the `CaseReducers` instead of the derived `CaseReducerActions`.

Then you have to create a new type that maps  the values of CaseReducers to always the reducer, and use that for the caseReducers.

Reply:
Another option with maybe even less changes would be to change your definition of 
```typescript
// from
type SliceCaseReducers<State, PA extends PayloadActions> = {
  [ActionType in keyof PA]: CaseReducer<State, PA[ActionType]>
}
// to something like
type SliceCaseReducers<State, ActionCreators extends  { [key: string]: unknown }> = {
  [ActionType in keyof ActionCreators]: CaseReducer<State, ReturnValue<ActionCreators[ActionType]>>
}
```

And actually pass ActionCreators as a second parameter to this interface, instead of any.

But honestly, I'm not sure if that will work out in the long run (also, I've written this without actually testing it, so it might not work)

Reply:
Trying your `CaseReducers` suggestion now, and I'm getting two errors:

- `Type 'unknown' does not satisfy the constraint 'Record<string, WithPayload<any, Action<string>>>'`
- `Required type parameters may not follow optional type parameters.`

I can resolve the first one by replacing `unknown` with `PayloadAction` (the record type).  Trying to figure out what an appropriate default type might be here.

Reply:
Resolved the second error by simply swapping the order of the generics in `Slice` :)

So. Coming up with a type for the reducers. I feel like the `CaseReducerActions` type is roughly what I want to mimic. I "just" need to figure out how to reference the reducer functions instead.

Reply:
And I _think_ I have something!